### ISSUE TITLE
Preemphasis generates a glitch in the start of the output signal

### ISSUE THREAD

**Issue by gerazov** [librosa/librosa#1032]
The current implementation of the preemphasis generates a glitch in the output signal.

Steps to reproduce:
```python
a = np.array([9.12e-05, 9.15e-05, 9.13e-05, 6.10e-05, 3.05e-05, 3.07e-05, 6.10e-05])
b = librosa.effects.preemphasis(a)
print(b)

Out: 
array([-9.699088e-01,  3.036000e-06,  2.545000e-06, -2.756100e-05,
       -2.867000e-05,  1.115000e-06,  3.122100e-05])
```
Here `b[0]` is clearly off.

What can be done is to asign `b[1]` to `b[0]`. This hack would deal with the glitch while keeping the lengths of `b` and `a` the same.

**Comment by bmcfee** [comment]
The filter response is undefined at the first sample. The default behavior is to initialize the filter to steady state, but if you want the behavior you describe, you can call preemphasis with `zi=[0]`.

**Comment by gerazov** [comment]
I see, but then `zi` should default to `[0]`, instead of `None`, at least for audio. Because this glitch hardly seems useful.

**Comment by lostanlen** [comment]
Hello @gerazov, thank you for reporting this problem.
This issue is the Doppelgänger of #851.

I could be convinced that `zi=[0]` is indeed a better default than `zi=None`.
This ties back to #902. Is there a use case for pre-emphasis on signals with nonzero average? (energy bands, for example).

In any case, changing the default would break backwards compatibility, so we cannot merge it before at least next year. (should we add a `FutureWarning` before changing this?)

**Comment by bmcfee** [comment]
> This ties back to #902. Is there a use case for pre-emphasis on signals with nonzero average? (energy bands, for example).

I'm having trouble thinking of a use-case like that which wouldn't be better served by other high-pass operations implemented elsewhere (eg pcen or delta).

OTOH, if you're applying pre-emphasis to short frames of audio, any given frame might have a nonzero average depending on how much low-frequency content there is.  Arguably, the whole point of preemph is to remove that low-frequency content.

Anyway, I'm also happy to reconsider the default parameter here (especially as it relates to #1025), but as @lostanlen says, any changes to default behavior should go through a deprecation cycle.

---
Unrelated comment: we should add some broadcast shaping magic to support `zi=scalar` as an argument.

**Comment by gerazov** [comment]
: ) Ok, I missed that one @lostanlen. In fact, #851 really describes how I discovered the problem with preemphasis. 

Here's what I got for the log spectrogram of a preemphasised audio signal with `zi = None`:
<img src="https://user-images.githubusercontent.com/15214418/70327777-d2600f00-1837-11ea-8958-8ea0682ec74d.png" width="300" />
And for comparison in the time domain (before and after):
<p float="left">
<img src="https://user-images.githubusercontent.com/15214418/70327898-1c48f500-1838-11ea-97cd-42c7e9d39ebe.png" width="250" /> 
<img src="https://user-images.githubusercontent.com/15214418/70327903-1fdc7c00-1838-11ea-8468-6f5e8c5ddcf4.png" width="250" />
</p>
This is clearly not what should happen. And a less informed user might use it as is without noticing the glitch.

I was thinking about @bmcfee's suggestion and I think `zi=[0]` is also not the best default value. For example we work with short vowel segments, which are not necessarily extracted at a zero crossing. Hence, the first value might be at the peak of the waveform, and a `zi = [0]` would again result with a glitch.

What I originally suggested would be equivalent to `reflect` padding in `np.pad` terms with `reflect_type='odd'`, i.e. `zi = - y[1]`. Other alternatives would be (again from `np.pad`):
- `reflect` (non odd) - `zi = y[1]`, or
- `symmetric` - `zi = y[0]` giving a zero at the start of the output signal.

In my opinion, any of these would be better than `zi = [0]`, and definitely better than the current `zi = None`.

**Comment by lostanlen** [comment]
@gerazov all this is good thinking. Thank you for shelling out the rationale here. I'm under the impression that the least amount of glitch would be produced by something along the lines of
`zi = 2*y[0] - y[1]`, as a form of linear extrapolation. But is this what we want?

**Comment by gerazov** [comment]
Well, I'm not a fan of extrapolation in audio, even if for one sample it shouldn't go very wrong :smile:  The default padding you use for `stft` is `reflect`, so in terms for consistency I vote for `zi = y[1]`.

**Comment by bmcfee** [comment]
Did we ever settle on a plan for this one?  I generally agree with @gerazov that extrapolation is bad times in general, but it seems pretty sensible in this specific case.

**Comment by lostanlen** [comment]
Yes, considering that values of `coef` are typically very close to `1`, i would still support linear extrapolation. Replicating `y[0]` ("constant") or `y[1]` ("reflect") would break the continuity of the first derivative, and thus create potential artifacts in the emphasized output.

**Comment by lostanlen** [comment]
@gerazov FYI, your issue is solved! (see #1159) will be included in 0.8
thanks again for reporting

**Comment by gerazov** [comment]
Awesome :sunglasses: keep up the good work guys!
